{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "46c6ebfd",
   "metadata": {},
   "source": [
    "Assignment 2 student 2 BINP16 HT25\n",
    "\n",
    "Author: Ariane Neumann\n",
    "\n",
    "Date: 2025-10-16\n",
    "\n",
    "Version:1.0\n",
    "\n",
    "Sytem: #!/usr/bin/envA python3 - coding: utf-8 -\n",
    "\n",
    "Description: This code will process a fasta file containing DNA sequences, translating them into protein sequences, count the abundances of the individual amino acids and give an output file. Different steps will be taken to achieve this:\n",
    "\n",
    "- Inspecting all available files\n",
    "- Creating a dictionary for the standard codons\n",
    "- Reading the files\n",
    "- Writing the script\n",
    "- Handle errors\n",
    "\n",
    "Functions created: generate_genetic_code, read_fasta, translate_DNA\n",
    "\n",
    "\n",
    "---------------------------------------------------------------------------------------------------------------------------------------"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "id": "b4a7e3b4-4d40-4125-be88-67a145037b20",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Preparation of the standard genetic code\n",
    "import sys\n",
    "import os\n",
    "\n",
    "if len(sys.argv) != 3:\n",
    "    sys.argv = [\"dna2aa.py\", \"DNA_seq.fna\", \"output_file.txt\"] \n",
    "DNA_seq = sys.argv[1]\n",
    "output_file = sys.argv[2]\n",
    "\n",
    "def generate_genetic_code():\n",
    "    bases = ['T', 'C', 'A', 'G']\n",
    "    # Making a dictionary with the codons (key) and the corresponding amino acid (value)\n",
    "    genetic_code = {'GTT': 'V', 'GTC': 'V', 'GTA': 'V', 'GTG': 'V',  # Valine\n",
    "    'GCT': 'A', 'GCC': 'A', 'GCA': 'A', 'GCG': 'A',  # Alanine\n",
    "    'GAT': 'D', 'GAC': 'D',  # Aspartic Acid\n",
    "    'GAA': 'E', 'GAG': 'E',  # Glutamic Acid\n",
    "    'GGT': 'G', 'GGC': 'G', 'GGA': 'G', 'GGG': 'G',  # Glycine\n",
    "    'TTT': 'F', 'TTC': 'F',  # Phenylalanine\n",
    "    'TTA': 'L', 'TTG': 'L',  # Leucine\n",
    "    'TCT': 'S', 'TCC': 'S', 'TCA': 'S', 'TCG': 'S',  # Serine\n",
    "    'TAT': 'Y', 'TAC': 'Y',  #Tyrosine\n",
    "    'TGT': 'C', 'TGC': 'C',  # Cysteine\n",
    "    'TGG': 'W',  # Tryptophan\n",
    "    'CTT': 'L', 'CTC': 'L', 'CTA': 'L', 'CTG': 'L',  # Leucine\n",
    "    'CCT': 'P', 'CCC': 'P', 'CCA': 'P', 'CCG': 'P',  # Proline\n",
    "    'CAT': 'H', 'CAC': 'H',  # Histidine\n",
    "    'CAA': 'Q', 'CAG': 'Q',  # Glutamine\n",
    "    'CGT': 'R', 'CGC': 'R', 'CGA': 'R', 'CGG': 'R',  # Arginine\n",
    "    'ATT': 'I', 'ATC': 'I', 'ATA': 'I', # Isoleucine\n",
    "    'ATG': 'M',  #  Methionine (Start)\n",
    "    'ACT': 'T', 'ACC': 'T', 'ACA': 'T', 'ACG': 'T',  # Threonine\n",
    "    'AAT': 'N', 'AAC': 'N',  # Asparagine\n",
    "    'AAA': 'K', 'AAG': 'K',  # Lysine\n",
    "    'AGT': 'S', 'AGC': 'S',  # Serine\n",
    "    'AGA': 'R', 'AGG': 'R',  # Arginine\n",
    "    'TAA': '*', 'TAG': '*', 'TGA': '*'} # Stop (*)\n",
    "\n",
    "#--------------------------------------------------------------------------------------------------------------------------\n",
    "\n",
    "# Define \"Reading the fasta file\" function\n",
    "def read_fasta(file_path):\n",
    "    sequences = {}\n",
    "    current_ID = None\n",
    "    current_sequence_lines = []\n",
    "    valid_DNA_char = {'A', 'C', 'T', 'G', 'N'} # Define a set of valid DNA characters for validation\n",
    "\n",
    "# Using 'with open' for secure file handling\n",
    "    try: \n",
    "        with open(file_path, 'r') as f:\n",
    "            line_number = 0\n",
    "            for line in f:\n",
    "                line_number += 1\n",
    "                line=line.strip() #removing whitespace\n",
    "                if line.startswith(\">\"):\n",
    "                    if current_ID:\n",
    "                        sequences[current_ID] = \"\".join(current_sequence_lines)\n",
    "                    current_ID=line[1:].strip()\n",
    "                    current_sequence_lines=[]\n",
    "                elif current_ID:\n",
    "                    clean_line=line.upper()\n",
    "                    for char in clean_line:\n",
    "                        if char not in valid_DNA_char:\n",
    "                            raise ValueError(f\" Found invalid character '{first_invalid_char}' in sequence {current_ID}. Only A, T, C, G, N are allowed.\" )\n",
    "                        current_sequence_line.append(clean_line)\n",
    "                        if current_ID and current_sequence_lines:\n",
    "                            sequences[current_ID]=\"\".join(current_sequence_lines)\n",
    "\n",
    "    except FileNotFoundError:\n",
    "        print(f\"Error:Input file not found at '{file_path}'\", file=sys.stderr)\n",
    "        return None\n",
    "    except ValueError:\n",
    "        print(f\"Error in file content\", file=sys.stderr)\n",
    "        return None\n",
    "    if not sequences:\n",
    "        print(\" Error: No sequences found in input file.\", file=sys.stderr)\n",
    "        return sequences\n",
    "        \n",
    "#--------------------------------------------------------------------------------------------------------------------------\"\n",
    "\n",
    "# Define \" Translate DNA\" function\n",
    "def translate_DNA (DNA_sequence, codon_table): #->str\n",
    "    aa_sequence=[] # empty list\n",
    "    sequence_length=len(DNA_sequence)\n",
    "    i=0\n",
    "\n",
    "    while i < sequence_length -2:\n",
    "        codon=DNA_sequence[i:i+3]\n",
    "\n",
    "        # need to check for base \"N\" (unknown base)\n",
    "        if \"N\" in codon:\n",
    "            aa_sequence.append(\"X\")\n",
    "        else:\n",
    "            amino_acid=codon_table.get(codon, \"X\")\n",
    "            aa_sequence.append(amino_acid)\n",
    "        i +=3\n",
    "    return \"\".join(aa_sequence)\n",
    "\n",
    "#---------------------------------------------------------------------------------------------------------------------------\n",
    "\n",
    "# Reading a fasta file in each line to read one a screen, the max length is set to 60\n",
    "maximum_line_length =60\n",
    "\n",
    "# Define \"execution of translation\" function\n",
    "def exec():\n",
    "    if len(sys.argv) <2:\n",
    "        print(f\"Usage: python3 {os.path.basename(sys.argv[0])} <input_DNA_fasta> [output_aa_fasta]\", file=sys.stderr)\n",
    "        sys.exit(1) #save and gentle exit\n",
    "\n",
    "    DNA_seq.fna=sys.argv[1]\n",
    "    output_file=sys.argv[2] if len(sys.argv) >2 else \"translated_aa.faa\"\n",
    "    print(f\"Start DNA translation from '{input_file}' to '{output_file}' ...\")\n",
    "\n",
    "    # Read fasta input file\n",
    "    dna_sequences=read_fasta(input_file)\n",
    "    if dna_sequences is None:\n",
    "        sys.exit(1) # in case no file is found, exit safely\n",
    "\n",
    "    # create an empty dictionary\n",
    "    translated_sequences={}\n",
    "    #translate the sequences with FOR loop\n",
    "    for seq_ID, DNA_seq in dna_sequences.items():\n",
    "        if len(DNA_seq) <3:\n",
    "            print(f\" Warning: Sequence '{seq_ID}' is too short for a translation (length: {len(DNA_seq)}). Skip this.\", file=sys.stderr)\n",
    "            continue\n",
    "\n",
    "        aa_seq=translate_dna(DNA_seq, genetic_code)\n",
    "        translated_sequences[seq_ID]=aa_seq\n",
    "\n",
    "        # Write an output file\n",
    "        output_successful = False\n",
    "\n",
    "    try:\n",
    "        with open(output_file,\"w\") as output_f:\n",
    "            for seq_ID, aa_seq in translated_sequences.items():\n",
    "                output_f.write(f\">{seq_ID}\\n\")\n",
    "                i=0\n",
    "                while i < len(aa_seq):\n",
    "                    output_f.write(aa_seq[i:i+ maximum_line_length] + '\\n')\n",
    "                    i += maximum_line_length\n",
    "            output_successful=True\n",
    "    except FileNotFoundError:\n",
    "        print(f\"Error: Input file not found at '{file_path}'\", file=sys.stderr)\n",
    "        return None\n",
    "    except ValueError:\n",
    "        print(f\"Error in file content\", file=sys.stderr)\n",
    "        return None      "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "07b1b998-1e9e-457d-bea9-4b733284ea0c",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.13.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
